docs/content/docs/concepts/glossary.md

```diff
@@ -25,182 +25,605 @@ under the License.
+partitioning methods such as shuffle, forward, and keyBy.
+
+#### Lateness
+
+Lateness is defined relative to the [watermarks](#watermark). A watermark(t) asserts that the stream 
+is complete up through to time t. Any event is considered late if it comes after the watermark whose 
+timestamp is ≤ t.
+
+#### ListState<T>
+
+This is a type of [keyed state](#keyed-state) that keeps a list of elements. You can append elements 
+and retrieve an Iterable over all currently stored elements. Elements are added using add(T) or 
+addAll(List<T>). The Iterable can be retrieved using Iterable<T> get().
 
 #### Logical Graph
 
-A logical graph is a directed graph where the nodes are  [Operators](#operator)
-and the edges define input/output-relationships of the operators and correspond
-to data streams or data sets. A logical graph is created by submitting jobs
-from a [Flink Application](#flink-application).
+This is a directed graph where the nodes are [operators](#operator) and the edges define input/output 
+relationships of the operators and correspond to [DataStreams](#datastreams). A logical graph is created 
+by submitting jobs to a [Flink cluster](#(flink)-cluster) from a [Flink application](#(flink)-application).
 
-Logical graphs are also often referred to as *dataflow graphs*.
+Logical graphs are also often referred to as [dataflow](#dataflow).
 
 #### Managed State
 
-Managed State describes application state which has been registered with the framework. For
-Managed State, Apache Flink will take care about persistence and rescaling among other things.
+Managed state is application state which has been registered with the stream processing framework, 
+which will take care of the persistence and rescaling of this state.  
+
+This type of state is represented in data structures controlled by the Flink runtime, such as internal 
+hash tables, or RocksDB. Flink’s runtime encodes the states and writes them into the checkpoints.
+
+[Keyed state](#keyed-state) and [operator state](#operator-state) exist in two forms: managed and [raw](#raw-state).
+
+#### MapState<UK, UV>
+
+This is a type of [keyed state](#keyed-state) that keeps a list of mappings. You can put key-value 
+pairs into the state and retrieve an Iterable over all currently stored mappings. Mappings are added 
+using put(UK, UV) or putAll(Map<UK, UV>). The value associated with a key can be retrieved using get(UK).
+
+#### Non-keyed State
+
+This type of state is bound to one parallel operator instance and is also called [operator state](#operator-state). 
+
+It is possible to work with [managed state](#managed-state) in non-keyed contexts but it is unusual 
+for user-defined functions to need non-keyed state and the interfaces involved would be different. 
+
+This feature is most often used in the implementation of [sources](#source) and [sinks](#sink).
+
+#### Offset
+
+A number identifying how far you are from the beginning of a certain [DataStream](#datastream). 
```

Comment:
The offset usually refers to a stream of data in the external syste, not a Flink DataStream.